src/main/scala/inox/ast/SymbolOps.scala

```diff
@@ -115,6 +116,50 @@ trait SymbolOps { self: TypeOps =>
     case _ => false
   }
 
+  /** Simplify all the pure and ground sub-expressions of the given expression,
+    * by evaluating them using [[inox.evaluators.Evaluator]].
+    * If `force` is omitted, the given expression will only be evaluated if it is
+    * ground, pure, and does not contain choose or quantifiers.
+    */
+  def simplifyGround(expr: Expr, force: Boolean = false)
+                    (implicit sem: symbols.Semantics, ctx: Context, opts: PurityOptions): Expr = {
+
+    val evalCtx = ctx.withOpts(evaluators.optEvalQuantifiers(false))
+    val evaluator = sem.getEvaluator(ctx)
+
+    case class SimplifyGroundError(expr: Expr, result: Result[Expr]) {
+      def errMsg: String = result match {
+        case RuntimeError(msg) =>
+          s"runtime error: $msg"
+        case EvaluatorError(msg) =>
+          s"evaluator error: $msg"
+        case _ => ""
+      }
+      override def toString: String = {
+        s"Forced evaluation of expression @ ${expr.getPos} failed because of a $errMsg"
+      }
+    }
+
+    def evalChildren(e: Expr): Expr = e match {
+      case Operator(es, recons) => recons(es.map(rec))
+    }
+
+    def rec(e: Expr): Expr = e match {
+      case e if isValue(e) =>
+        e
+      case e if force || isPure(e) && isGround(e) =>
```

Comment:
Doesn't `||` bind weaker than `&&` in Scala? Your expression needs to _at least_ be ground to be evaluatable.
Also, I'm not sure I mentioned this or not but I still believe `force` should be handled only by Stainless. I see two options for force:
1. As you implemented it, force everything _under_ a function with the `@force` annotation. This basically means calling `simplifyGround` with `force = true` on the body of that function.
2. Evaluate all ground calls to functions that _have_ the `@force` annotation (should be implemented entirely outside of Inox).

Maybe @jad-hamza has some insights about which one would be most useful for proofs.

Reply:
> Doesn't || bind weaker than && in Scala? Your expression needs to at least be ground to be evaluatable.

It does indeed but my thinking that it might sometimes be possible to evaluate a non-ground expression if eg. the free variable is located in a branch that is never entered. It's definitely an edge-case and I don't mind only evaluating ground expressions only.

> Also, I'm not sure I mentioned this or not but I still believe force should be handled only by Stainless.

It's likely, in which case I must have misunderstood you.

> I see two options for force:

> 1. As you implemented it, force everything under a function with the @force annotation. This basically means calling simplifyGround with force = true on the body of that function.
> 2. Evaluate all ground calls to functions that have the @force annotation (should be implemented entirely outside of Inox).
> Maybe @jad-hamza has some insights about which one would be most useful for proofs.

Ah good idea, the second one actually seems more useful to me. What do you think @jad-hamza ?

Reply:
Sounds good to me! I guess the second option gives you finer grained control, because you could always wrap the only expressions you want to force in a fresh function, annotated by @force (e.g. in a function named "force")

Reply:
Ah I hadn't though of the non-ground branch. I still wouldn't try to run an evaluator on a non-ground expression (especially the `CodeGen` one). You can always run this within a fixpoint if you want with `simplifyByConstructor` which should simplify all your branching trees.